### ISSUE TITLE
Issues with IGraphics contextual menu and right click/mousedown behaviour

### ISSUE THREAD

**Issue by AlexHarker** [iPlug2/iPlug2#636]
There are a few things that should be reviewed here:

- You can turn contextual menus off, but only globally, not per control, and multi-parameter controls will simply provide the contextual menu for the first parameter (or the one in the clicked area) which is not always ideal.
- Empty context menus still interrupt mousing if created whilst the right click is idea whilst holding the left button also
- MouseUps don't get classed when the right click is made whilst holding the left down also, which can leave controls in a drawn state that reflects the clicked/dragging state.
- Controls returning noValIdx when queried for their parameter with mouse position on mousedown are still treated like the first parameter should be used for contextual menus and AAX mousing, even though the return implies that is not the case

These are all fixable with a little reworking, but the correct behaviour first needs to be agreed in each case.

**Comment by olilarkin** [comment]
This begins to deal with the first two things

https://github.com/iPlug2/iPlug2/tree/graphics/contextmenu

**Comment by olilarkin** [comment]
also the 3rd thing

**Comment by AlexHarker** [comment]
This looks good, and is pretty much exactly what I would have done if I'd had the time.

The first two items appear to be dealt with (although on VST3 the right-click will interrupt mousing if the host has a contextual menu, unless you turn it off globally - however, that might be OK - it's certainly all better.

Remaining issues:

- I don't see how mouse ups are dealt with (as you imply above). Currently ReleaseMouseCapture() doesn't do mouses (and is a bit blunt as it always clears all touches which could be improved for cases where we only need to remove one). If we do add mouse ups we shouldn't call them if IGraphics is destructing as they could call through to pure virtual and we should do it prior to checking if the cursor is hidden.
- I'm unclear on what happens with a multiple param control - I see that this gets passed through to the contextual menu, but I can't tell yet what it'll actually do on VST3.

**Comment by AlexHarker** [comment]
On a muliple parameter control testing on reaper on VST3 the context control fails but the mousing doesn't fall through to the normal, so it interrupts other mousing.

**Comment by olilarkin** [comment]
hmmm it was working for me in IPlugInstrument control with IVMultiSliderControl... which one are you testing?

**Comment by AlexHarker** [comment]
What should I expect to happen? I'm testing my own plugin. Fall through should work on AU etc., but for VST3 there is an issue. When you say "it was working" can you clarify which bit?

Also worth noting that if you are dragging a control and then hit the right button on VST3 the contextual menu interrupts other mousing - which may or may not be desirable - not sure on that one.

**Comment by olilarkin** [comment]
yeah i see. don't know why i thought right click was working. lets discuss on slack
